File: android/app/src/main/kotlin/com/baseapp/android/query/ListFeedLoader.kt
package com.baseapp.android.query

import com.baseapp.android.store.ListPaginationStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

object ListFeedLoader {
    suspend fun <Page> fetchCurrentPage(
        pager: ListPaginationStore,
        withLoading: Boolean,
        resetPagination: Boolean,
        fetchPage: suspend (Int, String?) -> Page,
        applyPage: (Page, Int) -> Unit,
        onError: (Throwable, Int) -> Page?,
    ) {
        pager.setLoading(resetPagination = resetPagination, withLoading = withLoading)
        try {
            if (resetPagination) {
                pager.resetPaginationState()
            }

            val pageNumber = pager.currentPage
            try {
                val page = fetchPage(pageNumber, pager.afterCursor(pageNumber))
                applyPage(page, pageNumber)
            } catch (error: Exception) {
                onError(error, pageNumber)?.let { recoveryPage ->
                    applyPage(recoveryPage, pageNumber)
                }
            }
        } finally {
            pager.clearLoading(resetPagination = resetPagination)
        }
    }

    suspend fun <Page> loadNextPage(
        pager: ListPaginationStore,
        fetchPage: suspend (Int, String?) -> Page,
        applyPage: (Page, Int) -> Unit,
        onError: (Throwable, Int) -> Page?,
    ) {
        val nextPage = pager.nextPageNumberForLoading() ?: return
        pager.setLoading(resetPagination = false, withLoading = false)
        try {
            val page = fetchPage(nextPage, pager.afterCursor(nextPage))
            applyPage(page, nextPage)
        } catch (error: Exception) {
            onError(error, nextPage)?.let { recoveryPage ->
                applyPage(recoveryPage, nextPage)
            }
        } finally {
            pager.clearLoading(resetPagination = false)
        }
    }
}

/** Debounces rapid search-as-you-type refreshes into one trailing refresh. */
class DebouncedRefreshCoordinator(private val scope: CoroutineScope) {
    private var job: Job? = null

    fun schedule(delayMs: Long = 250, action: suspend () -> Unit) {
        job?.cancel()
        job = scope.launch {
            delay(delayMs)
            action()
        }
    }
}

File: ios/App/Query/ListFeedLoader.swift
import Foundation

@MainActor
enum ListFeedLoader {
  static func fetchCurrentPage<Page>(
    pager: ListPaginationStore,
    withLoading: Bool,
    resetPagination: Bool,
    fetchPage: (Int, String?) async throws -> Page,
    applyPage: (Page, Int) -> Void,
    onError: (Error, Int) -> Page?
  ) async {
    pager.setLoading(resetPagination: resetPagination, withLoading: withLoading)
    defer {
      pager.clearLoading(resetPagination: resetPagination)
    }

    if resetPagination {
      pager.resetPaginationState()
    }

    let pageNumber = pager.currentPage
    do {
      let page = try await fetchPage(pageNumber, pager.afterCursor(forPage: pageNumber))
      applyPage(page, pageNumber)
    } catch {
      if let recoveryPage = onError(error, pageNumber) {
        applyPage(recoveryPage, pageNumber)
      }
    }
  }

  static func loadNextPage<Page>(
    pager: ListPaginationStore,
    fetchPage: (Int, String?) async throws -> Page,
    applyPage: (Page, Int) -> Void,
    onError: (Error, Int) -> Page?
  ) async {
    guard let nextPage = pager.nextPageNumberForLoading() else {
      return
    }
    await fetchSpecificPageWithPageLoadingState(
      pager: pager,
      pageNumber: nextPage,
      fetchPage: fetchPage,
      applyPage: applyPage,
      onError: onError
    )
  }

  private static func fetchSpecificPageWithPageLoadingState<Page>(
    pager: ListPaginationStore,
    pageNumber: Int,
    fetchPage: (Int, String?) async throws -> Page,
    applyPage: (Page, Int) -> Void,
    onError: (Error, Int) -> Page?
  ) async {
    pager.setLoading(resetPagination: false, withLoading: false)
    defer {
      pager.clearLoading(resetPagination: false)
    }

    do {
      let page = try await fetchPage(pageNumber, pager.afterCursor(forPage: pageNumber))
      applyPage(page, pageNumber)
    } catch {
      if let recoveryPage = onError(error, pageNumber) {
        applyPage(recoveryPage, pageNumber)
      }
    }
  }
}

@MainActor
final class DebouncedRefreshCoordinator {
  private var task: Task<Void, Never>?

  func schedule(
    delayNanoseconds: UInt64 = 250_000_000,
    action: @escaping @MainActor () async -> Void
  ) {
    task?.cancel()
    task = Task {
      try? await Task.sleep(nanoseconds: delayNanoseconds)
      guard !Task.isCancelled else {
        return
      }
      await action()
    }
  }

  deinit {
    task?.cancel()
  }
}
